File: Editorial/2001-2100/2083_Substrings_That_Begin_and_End_With_the_Same_Letter/Python3/2_Counting.py
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        answer = 0
        frequency_count = [0] * 26

        # Count the frequency of each character in the string.
        for ch in s:
            frequency_count[ord(ch) - ord("a")] += 1

        # Calculate the total number of valid substrings.
        for current_count in frequency_count:
            # Using (current_count + 1) choose 2 to calculate valid substrings
            # for the current letter.
            answer += ((current_count + 1) * current_count) // 2

        return answer

File: Editorial/2001-2100/2083_Substrings_That_Begin_and_End_With_the_Same_Letter/CPP/2_Counting.cpp
class Solution {
public:
    long long numberOfSubstrings(string s) {
        long long answer = 0;
        vector<long long> frequencyCount(26, 0);

        // Count the frequency of each character in the string.
        for (char ch : s) {
            frequencyCount[ch - 'a']++;
        }

        // Calculate the total number of valid substrings.
        for (long long currentCount : frequencyCount) {
            // Using (currentCount + 1) choose 2 to calculate valid substrings
            // for the current letter.
            answer += ((currentCount + 1) * currentCount) / 2;
        }

        return answer;
    }
};
